lisa/tests/scheduler/numa_behaviour.py

```diff
@@ -0,0 +1,121 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (C) 2019, Linaro and contributors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+import os.path
+
+import pandas as pd
+
+from lisa.wlgen.rta import Periodic
+from lisa.tests.base import ResultBundle, RTATestBundle, TestMetric
+from lisa.utils import ArtifactPath
+from lisa.trace import requires_events, FtraceCollector
+from lisa.target import Target
+from lisa.datautils import df_filter_task_ids, df_deduplicate
+
+class NUMABehaviour(RTATestBundle):
+    """
+    Abstract class for NUMA related scheduler testing.
+
+    This class provides :meth:`test_task_placement` to validate the basic
+    behaviour of NUMA related scheduling decisions. 
+    """
+    @classmethod
+    def check_from_target(cls, target):
+        if target.number_of_nodes < 2:
+            raise CannotCreateError(
+                "Target doesn't have at least two NUMA nodes")
+
+    @requires_events('sched_switch')
+    def _get_task_cpu_df(self, task_id):
+        """
+        Get a DataFrame for task mirgrations
+
+        Use the sched_switch trace event to find task mirgation from one CPU to another.
+
+        :returns: A Pandas DataFrame for the task, showing the
+                  CPU's that the task was migrated to
+        """        
+        df = self.trace.df_events('sched_switch')[['next_pid', 'next_comm', '__cpu']]
+        cpu_df = df_filter_task_ids(df, [task_id], pid_col='next_pid', comm_col='next_comm')
```

Comment:
IMO those first two lines should be reduced to:
```
df = self.trace.analysis.tasks.df_task_states(task_id)
```

This keeps the logic of which columns to get / which column to filter on in the above helper.
You can then call `df_deduplicate` with `cols=['cpu']`.

Reply:
Sure, will do